File: aoc2020/src/day7/part1.ts
type BagRule = {
  color: string;
  count: number;
};

function parseRules(rules: string): Map<string, BagRule[]> {
  const bagRules = new Map<string, BagRule[]>();
  const ruleLines = rules.trim().split("\n");

  for (const rule of ruleLines) {
    const [containerColor, contents] = rule.split(" contain ");
    const containerColorParts = containerColor.split(" ").slice(0, -1);
    const containerColorKey = containerColorParts.join(" ");

    if (contents === "no other bags.") {
      continue;
    }

    const contentRules: BagRule[] = contents
      .slice(0, -1)
      .split(", ")
      .map((item) => {
        const [count, color] = item.split("", 2);
        const colorParts = color.split(" ").slice(0, -1);
        const colorKey = colorParts.join(" ");
        return { color: colorKey, count: parseInt(count, 10) };
      });
    bagRules.set(containerColorKey, contentRules);
  }

  return bagRules;
}

function countContainers(rules: string, startBag: string): number {
  const bagRules = parseRules(rules);
  const queue: string[] = [startBag];
  const containers = new Set<string>();

  while (queue.length > 0) {
    const currBag = queue.shift();
    for (const [containerColor, contentRules] of bagRules.entries()) {
      if (contentRules.some((rule) => rule.color === currBag)) {
        containers.add(containerColor);
        queue.push(containerColor);
      }
    }
  }

  return containers.size;
}

async function main() {
  const file = Bun.file("./input.txt");
  const arrBuffer = await file.arrayBuffer();
  const buffer = Buffer.from(arrBuffer);
  const rules = buffer.toString();
  console.log(countContainers(rules,"shiny gold"));
}

await main();

File: 2020/day7/part1.py
from typing import Dict, List
from dataclasses import dataclass

@dataclass
class BagRule:
    color: str
    count: int

def parse_rules(rules: str) -> Dict[str, List[BagRule]]:
    bag_rules = {}
    rule_lines = rules.strip().split("\n")

    for rule in rule_lines:
        container_color, contents = rule.split(" contain ")
        container_color_parts = container_color.split()[:-1]  # Remove 'bags'/'bag'
        container_color_key = " ".join(container_color_parts)

        if contents == "no other bags.":
            continue

        content_rules = []
        for item in contents[:-1].split(", "):  # Remove period and split
            words = item.split()
            count = int(words[0])
            color_parts = words[1:-1]  # Remove 'bags'/'bag'
            color_key = " ".join(color_parts)
            content_rules.append(BagRule(color=color_key, count=count))

        bag_rules[container_color_key] = content_rules

    return bag_rules

def count_containers(rules: str, start_bag: str) -> int:
    bag_rules = parse_rules(rules)
    queue = [start_bag]
    containers = set()

    while queue:
        curr_bag = queue.pop(0)
        for container_color, content_rules in bag_rules.items():
            if any(rule.color == curr_bag for rule in content_rules):
                containers.add(container_color)
                queue.append(container_color)

    return len(containers)

def main():
    with open("input.txt", "r") as file:
        rules = file.read()
    print(count_containers(rules, "shiny gold"))

if __name__ == "__main__":
    main()
